<corecrt.h> redefinition warnings and assert.h(17,1) syntax error

### Environment

```toml
[System]
  name    = "Windows 11 Pro"
  version = "23H2"
  build   = "22631.4169"

[Compiler]
  MSVC           = "17 2022"
  version        = "19.40.33813"
  toolset        = "v143"
  system_version = "10.0.22631"

[Build System]
  CMake        = "3.30.3"
  CXX_FLAGS    = "/MP12 /EHsc /Oi"
  cpp_standard = 17
  c_standard   = 11
  architecture = "x64"

[wxWidgets]
  version = "3.3.0"
  port    = "wxMSW"
```

### Description

I'm trying to build and link the `wxWidgets` library as the subdirectory:

```CMake
set(wxBUILD_COMPATIBILITY 3.2)
set(wxBUILD_PRECOMP       OFF)
set(wxBUILD_SHARED        OFF)
set(wxBUILD_MONOLITHIC    OFF)

add_subdirectory(${CD_DEPS_DIR}/wxWidgets "editor_deps/wxWidgets" EXCLUDE_FROM_ALL)
target_link_libraries(${target} PRIVATE wx::core wx::base wx::gl)
```

And I get the following warnings twice:

```build
[build] C:\Program Files (x86)\Windows Kits\10\Include\10.0.26100.0\um\commctrl.h(205,9): warning C4005: 'CCM_SETUNICODEFORMAT': macro redefinition [G:\proj\build\msvc_editor\code\editor\editor_deps\wxWidgets\libs\core\wxcore.vcxproj]
[build] C:\Program Files (x86)\Windows Kits\10\Include\10.0.26100.0\um\commctrl.h(714,9): warning C4005: 'HDS_HOTTRACK': macro redefinition [G:\proj\build\msvc_editor\code\editor\editor_deps\wxWidgets\libs\core\wxcore.vcxproj]
[build] C:\Program Files (x86)\Windows Kits\10\Include\10.0.26100.0\um\commctrl.h(2982,9): warning C4005: 'CCS_RIGHT': macro redefinition [G:\proj\build\msvc_editor\code\editor\editor_deps\wxWidgets\libs\core\wxcore.vcxproj]
```

And those errors:

```log
[build] C:\Program Files (x86)\Windows Kits\10\Include\10.0.26100.0\ucrt\assert.h(17,1): error C2059: syntax error: 'extern ' [G:\proj\build\msvc_editor\cd2.vcxproj]
[build] C:\Program Files (x86)\Windows Kits\10\Include\10.0.26100.0\ucrt\assert.h(17,1): error C2143: syntax error: missing ';' before '{' [G:\proj\build\msvc_editor\cd2.vcxproj]
[build] C:\Program Files (x86)\Windows Kits\10\Include\10.0.26100.0\ucrt\assert.h(17,1): error C2447: '{': missing function header (old-style formal list?) [G:\proj\build\msvc_editor\cd2.vcxproj]
```

---

Seems like the `"wx/msw/missing.h"` file somehow includes just before the `<commctrl.h>`, but I cannot get how would it be.

`dialog.cpp` is including `wrapcdlg.h`, which then includes `missing.h`. Here is the first problem, `missing.h` should be included after `<commdlg.h>` so it defines the stuff that wasn't in the Windows header.

But this isn't the only problem, because later on, `dialog.cpp` is including `wrapcctl.h`, which includes `<commctrl.h>`. But `missing.h` already defined some stuff from `<commctrl.h>` so this is causing the warnings.

In this case including `wrapcctl.h` before `wrapcdlg.h` fixes the problem, and I'll create a PR doing this.

But in theory it could be possible that this causes again warnings because now `missing.h` is included (via `wrapcctl.h`) but it defines stuff that is later included via `wrapcdlg.h`. I thought of some solutions but none of them are ideal.
We could define stuff inside `missing.h` inside `_INC_COMMCTRL`, `_INC_COMMDLG` and other guards. But Microsoft could decide to change these names. And they might be different in MinGW headers.
We could include all used Windows headers, and then include `missing.h` but this will probably make non-precompiled builds even slower.
Or use separate `missing_commctrl.h`, `missing_commdlg.h` files for each Windows header, or move the missing defines into the wrapper. But then everything is distributed over many files.
So for now I'll just fix this problem and hope it will be fine forever.

I have no idea where the errors come from, but this is likely something in your own code.
The list of includes also shows
```
[build] G:\proj\code\editor\editor\frames\mainFrame.cpp(63,1): error C1004: unexpected end-of-file found [G:\proj\build\msvc_editor\cd2.vcxproj]
```

--- a/build/cmake/options.cmake
+++ b/build/cmake/options.cmake
@@ -392,7 +392,6 @@ wx_option(wxUSE_TREELISTCTRL "use wxTreeListCtrl class")
 # common dialogs
 # ---------------------------------------------------------------------------
 
-wx_option(wxUSE_COMMON_DIALOGS "use all common dialogs")
 wx_option(wxUSE_ABOUTDLG "use wxAboutBox")
 wx_option(wxUSE_CHOICEDLG "use wxChoiceDialog")
 wx_option(wxUSE_COLOURDLG "use wxColourDialog")

--- a/build/cmake/setup.h.in
+++ b/build/cmake/setup.h.in
@@ -445,8 +445,6 @@
 #endif
 
 
-#cmakedefine01 wxUSE_COMMON_DIALOGS
-
 #cmakedefine01 wxUSE_BUSYINFO
 
 #cmakedefine01 wxUSE_CHOICEDLG

--- a/docs/doxygen/mainpages/const_wxusedef.h
+++ b/docs/doxygen/mainpages/const_wxusedef.h
@@ -308,7 +308,6 @@ library:
 @itemdef{wxUSE_ACCESSIBILITY, Enable accessibility support}
 @itemdef{wxUSE_ACTIVEX, Use wxActiveXContainer and related classes.}
 @itemdef{wxUSE_COMBOCTRL_POPUP_ANIMATION, See wx/msw/combo.h file.}
-@itemdef{wxUSE_COMMON_DIALOGS, Enable use of windows common dialogs from header commdlg.h; example PRINTDLG.}
 @itemdef{wxUSE_CRASHREPORT, Use wxCrashReport class.}
 @itemdef{wxUSE_DATEPICKCTRL_GENERIC, Use generic wxDatePickerCtrl implementation in addition to the native one.}
 @itemdef{wxUSE_DC_CACHEING, cache temporary wxDC objects.}

--- a/include/wx/android/setup.h
+++ b/include/wx/android/setup.h
@@ -1184,14 +1184,6 @@
 // common dialogs
 // ----------------------------------------------------------------------------
 
-// Use common dialogs (e.g. file selector, printer dialog). Switching this off
-// also switches off the printing architecture and interactive wxPrinterDC.
-//
-// Default is 1
-//
-// Recommended setting: 1
-#define wxUSE_COMMON_DIALOGS 1
-
 // wxBusyInfo displays window with message when app is busy. Works in same way
 // as wxBusyCursor
 #define wxUSE_BUSYINFO      1

--- a/include/wx/gtk/setup.h
+++ b/include/wx/gtk/setup.h
@@ -1184,14 +1184,6 @@
 // common dialogs
 // ----------------------------------------------------------------------------
 
-// Use common dialogs (e.g. file selector, printer dialog). Switching this off
-// also switches off the printing architecture and interactive wxPrinterDC.
-//
-// Default is 1
-//
-// Recommended setting: 1
-#define wxUSE_COMMON_DIALOGS 1
-
 // wxBusyInfo displays window with message when app is busy. Works in same way
 // as wxBusyCursor
 #define wxUSE_BUSYINFO      1

--- a/include/wx/msw/setup.h
+++ b/include/wx/msw/setup.h
@@ -1184,14 +1184,6 @@
 // common dialogs
 // ----------------------------------------------------------------------------
 
-// Use common dialogs (e.g. file selector, printer dialog). Switching this off
-// also switches off the printing architecture and interactive wxPrinterDC.
-//
-// Default is 1
-//
-// Recommended setting: 1
-#define wxUSE_COMMON_DIALOGS 1
-
 // wxBusyInfo displays window with message when app is busy. Works in same way
 // as wxBusyCursor
 #define wxUSE_BUSYINFO      1

--- a/include/wx/msw/wrapcdlg.h
+++ b/include/wx/msw/wrapcdlg.h
@@ -14,12 +14,10 @@
 
 #include "wx/msw/wrapwin.h"
 #include "wx/msw/private.h"
+
+#include <commdlg.h>
+
 #include "wx/msw/missing.h"
-
-#if wxUSE_COMMON_DIALOGS
-    #include <commdlg.h>
-#endif
-
 #include "wx/msw/winundef.h"
 
 #endif // _WX_MSW_WRAPCDLG_H_

--- a/include/wx/osx/setup.h
+++ b/include/wx/osx/setup.h
@@ -1191,14 +1191,6 @@
 // common dialogs
 // ----------------------------------------------------------------------------
 
-// Use common dialogs (e.g. file selector, printer dialog). Switching this off
-// also switches off the printing architecture and interactive wxPrinterDC.
-//
-// Default is 1
-//
-// Recommended setting: 1
-#define wxUSE_COMMON_DIALOGS 1
-
 // wxBusyInfo displays window with message when app is busy. Works in same way
 // as wxBusyCursor
 #define wxUSE_BUSYINFO      1

--- a/include/wx/setup_inc.h
+++ b/include/wx/setup_inc.h
@@ -1180,14 +1180,6 @@
 // common dialogs
 // ----------------------------------------------------------------------------
 
-// Use common dialogs (e.g. file selector, printer dialog). Switching this off
-// also switches off the printing architecture and interactive wxPrinterDC.
-//
-// Default is 1
-//
-// Recommended setting: 1
-#define wxUSE_COMMON_DIALOGS 1
-
 // wxBusyInfo displays window with message when app is busy. Works in same way
 // as wxBusyCursor
 #define wxUSE_BUSYINFO      1

--- a/include/wx/univ/setup.h
+++ b/include/wx/univ/setup.h
@@ -1184,14 +1184,6 @@
 // common dialogs
 // ----------------------------------------------------------------------------
 
-// Use common dialogs (e.g. file selector, printer dialog). Switching this off
-// also switches off the printing architecture and interactive wxPrinterDC.
-//
-// Default is 1
-//
-// Recommended setting: 1
-#define wxUSE_COMMON_DIALOGS 1
-
 // wxBusyInfo displays window with message when app is busy. Works in same way
 // as wxBusyCursor
 #define wxUSE_BUSYINFO      1

--- a/setup.h.in
+++ b/setup.h.in
@@ -448,8 +448,6 @@
 #endif
 
 
-#define wxUSE_COMMON_DIALOGS 0
-
 #define wxUSE_BUSYINFO      0
 
 #define wxUSE_CHOICEDLG     0

--- a/setup.h_vms
+++ b/setup.h_vms
@@ -501,8 +501,6 @@ typedef pid_t GPid;
 #define wxUSE_AUTOID_MANAGEMENT 0
 #endif
 
-#define wxUSE_COMMON_DIALOGS 1
-
 #define wxUSE_BUSYINFO      1
 
 #define wxUSE_CHOICEDLG     1

--- a/src/common/cmndata.cpp
+++ b/src/common/cmndata.cpp
@@ -24,9 +24,6 @@
 #include "wx/cmndata.h"
 
 #ifndef WX_PRECOMP
-    #if defined(__WXMSW__)
-        #include "wx/msw/wrapcdlg.h"
-    #endif // MSW
     #include <stdio.h>
     #include "wx/string.h"
     #include "wx/utils.h"

--- a/src/common/prntbase.cpp
+++ b/src/common/prntbase.cpp
@@ -16,9 +16,6 @@
 #include "wx/dcprint.h"
 
 #ifndef WX_PRECOMP
-    #if defined(__WXMSW__)
-        #include "wx/msw/wrapcdlg.h"
-    #endif // MSW
     #include "wx/utils.h"
     #include "wx/dc.h"
     #include "wx/app.h"
@@ -63,12 +60,6 @@
 #include "wx/generic/prntdlgg.h"
 #include "wx/dcps.h"
 #endif
-
-#ifdef __WXMSW__
-    #ifndef __WIN32__
-        #include <print.h>
-    #endif
-#endif // __WXMSW__
 
 // The value traditionally used as the default max page number and meaning
 // "infinitely many". It should probably be documented and exposed, but for now

--- a/src/msw/dc.cpp
+++ b/src/msw/dc.cpp
@@ -20,7 +20,7 @@
 
 
 #ifndef WX_PRECOMP
-    #include "wx/msw/wrapcdlg.h"
+    #include "wx/msw/private.h"
     #include "wx/image.h"
     #include "wx/window.h"
     #include "wx/utils.h"

--- a/src/msw/dcprint.cpp
+++ b/src/msw/dcprint.cpp
@@ -74,69 +74,6 @@ wxIMPLEMENT_ABSTRACT_CLASS(wxPrinterDCImpl, wxMSWDCImpl);
 // wxPrinterDC construction
 // ----------------------------------------------------------------------------
 
-#if 0
-// This form is deprecated
-wxPrinterDC::wxPrinterDC(const wxString& driver_name,
-                         const wxString& device_name,
-                         const wxString& file,
-                         bool interactive,
-                         wxPrintOrientation orientation)
-{
-    m_isInteractive = interactive;
-
-    if ( !file.empty() )
-        m_printData.SetFilename(file);
-
-#if wxUSE_COMMON_DIALOGS
-    if ( interactive )
-    {
-        PRINTDLG pd;
-
-        pd.lStructSize = sizeof( PRINTDLG );
-        pd.hwndOwner = (HWND) nullptr;
-        pd.hDevMode = (HANDLE)nullptr;
-        pd.hDevNames = (HANDLE)nullptr;
-        pd.Flags = PD_RETURNDC | PD_NOSELECTION | PD_NOPAGENUMS;
-        pd.nFromPage = 0;
-        pd.nToPage = 0;
-        pd.nMinPage = 0;
-        pd.nMaxPage = 0;
-        pd.nCopies = 1;
-        pd.hInstance = (HINSTANCE)nullptr;
-
-        m_ok = PrintDlg( &pd ) != 0;
-        if ( m_ok )
-        {
-            m_hDC = (WXHDC) pd.hDC;
-        }
-    }
-    else
-#endif // wxUSE_COMMON_DIALOGS
-    {
-        if ( !driver_name.empty() && !device_name.empty() && !file.empty() )
-        {
-            m_hDC = (WXHDC) CreateDC(driver_name.t_str(),
-                                     device_name.t_str(),
-                                     file.fn_str(),
-                                     nullptr);
-        }
-        else // we don't have all parameters, ask the user
-        {
-            wxPrintData printData;
-            printData.SetOrientation(orientation);
-            m_hDC = wxGetPrinterDC(printData);
-        }
-
-        m_ok = m_hDC ? true: false;
-
-        // as we created it, we must delete it as well
-        m_bOwnsDC = true;
-    }
-
-    Init();
-}
-#endif
-
 wxPrinterDCImpl::wxPrinterDCImpl( wxPrinterDC *owner, const wxPrintData& printData ) :
     wxMSWDCImpl( owner )
     , m_printData(printData)

--- a/src/msw/dialog.cpp
+++ b/src/msw/dialog.cpp
@@ -23,7 +23,6 @@
 #include "wx/modalhook.h"
 
 #ifndef WX_PRECOMP
-    #include "wx/msw/wrapcdlg.h"
     #include "wx/utils.h"
     #include "wx/frame.h"
     #include "wx/app.h"

--- a/src/msw/filedlg.cpp
+++ b/src/msw/filedlg.cpp
@@ -24,8 +24,6 @@
 #include "wx/filedlg.h"
 
 #ifndef WX_PRECOMP
-    #include "wx/msw/wrapcdlg.h"
-    #include "wx/msw/missing.h"
     #include "wx/utils.h"
     #include "wx/msgdlg.h"
     #include "wx/filefn.h"
@@ -44,6 +42,8 @@
 #include "wx/tokenzr.h"
 #include "wx/modalhook.h"
 
+#include "wx/msw/wrapshl.h"
+#include "wx/msw/wrapcdlg.h"
 #include "wx/msw/private/dpiaware.h"
 #include "wx/msw/private/filedialog.h"
 
@@ -59,8 +59,6 @@
     #include "wx/radiobut.h"
     #include "wx/stattext.h"
     #include "wx/textctrl.h"
-
-    #include "wx/msw/wrapshl.h"
 
     #include "wx/msw/private/cotaskmemptr.h"
 #endif // wxUSE_IFILEOPENDIALOG

--- a/src/msw/graphics.cpp
+++ b/src/msw/graphics.cpp
@@ -15,7 +15,7 @@
 #if wxUSE_GRAPHICS_GDIPLUS
 
 #ifndef WX_PRECOMP
-    #include "wx/msw/wrapcdlg.h"
+    #include "wx/msw/private.h"
     #include "wx/image.h"
     #include "wx/window.h"
     #include "wx/utils.h"
@@ -42,11 +42,6 @@
 #include "wx/dcgraph.h"
 #include "wx/rawbmp.h"
 
-#include "wx/msw/private.h" // needs to be before #include <commdlg.h>
-
-#if wxUSE_COMMON_DIALOGS
-#include <commdlg.h>
-#endif
 #include <float.h> // for FLT_MAX, FLT_MIN
 
 #ifdef _MSC_VER

--- a/src/msw/printwin.cpp
+++ b/src/msw/printwin.cpp
@@ -24,7 +24,6 @@
 #if wxUSE_PRINTING_ARCHITECTURE && (!defined(__WXUNIVERSAL__) || !wxUSE_POSTSCRIPT_ARCHITECTURE_IN_MSW)
 
 #ifndef WX_PRECOMP
-    #include "wx/msw/wrapcdlg.h"
     #include "wx/window.h"
     #include "wx/msw/private.h"
     #include "wx/utils.h"

